Add support for user enforced spark property overrides in Profiling Tool's AutoTuner

parthosa opened the pull request:
Fixes #1526 

This MR adds support for user-enforced Spark property overrides in AutoTuner. This is required in cases where users have more context about their workloads and may wish to override AutoTuner’s default recommendations, which might not always be optimal.

The changes in this PR handle the interaction between the user enforced properties and how it would affect the recommendation logic for other properties or how it would affect the recommended cluster shape. 

For example: 
- User enforced value of memoryOverhead can be too large that might exceed the available memory limits. 
    - See UTs added in `ProfilingAutoTunerSuiteV2` for these cases.
- User enforced value of `spark.executor.instances` might affect the recommended cluster shape. 
    - See UTs added in `ClusterRecommendationSuite` for these cases.
 

### Usage via Python CLI
1. Build and install the Python CLI in non-fat mode

```shell
$ cd spark-rapids-tools/user_tools
$ ./build.sh non-fat
$ pip install dist/spark_rapids_user_tools-25.4.2-py3-none-any.whl
```

2. Create a target cluster YAML file
```yaml
workerInfo:
  instanceType: g2-standard-8
sparkProperties:
  enforced:
    spark.sql.shuffle.partitions: 400
    spark.sql.files.maxPartitionBytes: 100m
    spark.task.resource.gpu.amount: 0.25
    spark.rapids.sql.concurrentGpuTasks: 2
```

4. Run the Profile CLI as:
```shell
spark_rapids profiling \
  --platform $PLATFORM \
  --eventlogs $EVENT_LOG \
  --target-cluster-info $PATH_TO_TARGET_YAML \
```

5. Corresponding AutoTuner recommendations will generated at

<details>
<summary>File: prof_2025xxxx/profiling_summary.log</summary>
<pre>
### Recommended configurations ###
application_1746596730921_0001
	Spark Properties:
	--conf spark.dataproc.enhanced.execution.enabled=false
	--conf spark.dataproc.enhanced.optimizer.enabled=false
	--conf spark.executor.cores=8
	--conf spark.executor.instances=300
	--conf spark.executor.memory=16g
	--conf spark.executor.memoryOverhead=9830m
	--conf spark.locality.wait=0
	--conf spark.rapids.memory.pinnedPool.size=4g
	--conf spark.rapids.shuffle.multiThreaded.reader.threads=20
	--conf spark.rapids.shuffle.multiThreaded.writer.threads=20
	--conf spark.rapids.sql.batchSizeBytes=2147483647b
	--conf spark.rapids.sql.concurrentGpuTasks=2
	--conf spark.rapids.sql.enabled=true
	--conf spark.rapids.sql.multiThreadedRead.numThreads=40
	--conf spark.shuffle.manager=com.nvidia.spark.rapids.spark353.RapidsShuffleManager
	--conf spark.sql.adaptive.autoBroadcastJoinThreshold=[FILL_IN_VALUE]
	--conf spark.sql.adaptive.coalescePartitions.minPartitionSize=4m
	--conf spark.sql.adaptive.coalescePartitions.parallelismFirst=false
	--conf spark.sql.files.maxPartitionBytes=100m
	--conf spark.sql.shuffle.partitions=400
	--conf spark.task.resource.gpu.amount=0.25
	Comments:
	- 'spark.dataproc.enhanced.execution.enabled' should be disabled. WARN: Turning this property on might case the GPU accelerated Dataproc cluster to hang.
	- 'spark.dataproc.enhanced.execution.enabled' was not set.
	- 'spark.dataproc.enhanced.optimizer.enabled' should be disabled. WARN: Turning this property on might case the GPU accelerated Dataproc cluster to hang.
	- 'spark.dataproc.enhanced.optimizer.enabled' was not set.
	- 'spark.executor.memoryOverhead' was not set.
	- 'spark.rapids.memory.pinnedPool.size' was not set.
	- 'spark.rapids.shuffle.multiThreaded.reader.threads' was not set.
	- 'spark.rapids.shuffle.multiThreaded.writer.threads' was not set.
	- 'spark.rapids.sql.batchSizeBytes' was not set.
	- 'spark.rapids.sql.concurrentGpuTasks' was user-enforced in the target cluster properties.
	- 'spark.rapids.sql.enabled' was not set.
	- 'spark.rapids.sql.multiThreadedRead.numThreads' was not set.
	- 'spark.shuffle.manager' was not set.
	- 'spark.sql.adaptive.autoBroadcastJoinThreshold' was not set.
	- 'spark.sql.adaptive.enabled' should be enabled for better performance.
	- 'spark.sql.files.maxPartitionBytes' was user-enforced in the target cluster properties.
	- 'spark.sql.shuffle.partitions' was user-enforced in the target cluster properties.
	- 'spark.task.resource.gpu.amount' was user-enforced in the target cluster properties.
	- A newer RAPIDS Accelerator for Apache Spark plugin is available:
	https://repo1.maven.org/maven2/com/nvidia/rapids-4-spark_2.12/25.04.0/rapids-4-spark_2.12-25.04.0.jar
	Version used in application is 24.12.0.
	- The RAPIDS Shuffle Manager requires spark.driver.extraClassPath
	and spark.executor.extraClassPath settings to include the
	path to the Spark RAPIDS plugin jar.
	If the Spark RAPIDS jar is being bundled with your Spark
	distribution, this step is not needed.
</pre>
</details>

## Changes

- `TargetClusterProps.scala`
  - Added support for a new `sparkProperties` section.
  - Handled user-enforced Spark properties through the `enforced` field.
- `Platform.scala`
  - Introduced `getUserEnforcedSparkProperty()` to retrieve user-overridden Spark configs.
  - Refactored property access to use a function `sparkPropertiesFn`, (which prioritizes user-enforced values over source values)
- `AutoTuner.scala`
  - <ins> Updated `calcOverallMemory()` to account for user-enforced Spark properties when computing memory-related recommendations. </ins>
    - For example, user-provided memory overhead or pinned memory pool sizes may now exceed container limits.
    -  Unit tests covering these scenarios added to `ProfilingAutoTunerSuiteV2`.
  - Added `getAllSourceProperties` and `getPropertyValueFromSource()` to clearly distinguish between source and enforced property origins.
  - For all user-enforced properties, AutoTuner would include a comment as
     - `'spark.task.resource.gpu.amount' was user-enforced in the target cluster properties.`
- `ClusterConfigurationStrategy.scala`
  - <ins>Integrated user-enforced values like `spark.executor.cores` and `spark.executor.instances` into cluster configuration logic</ins>
    - For example, `spark.executor.instances` affect number of GPUs and thus the number of worker nodes.
    - Unit tests covering these scenarios added to `ClusterRecommendationSuite`.

## Tests
- Added unit tests for tuning overrides in `ProfilingAutoTunerSuiteV2`.
- Added unit tests validating the interaction between tuning overrides and cluster shape in `ClusterRecommendationSuite`.
- Added a note in `ProfilingAutoTunerSuite` indicating that no new AutoTuner tests should be added to this class
   - It uses the legacy worker info properties format, which is overloaded to be used for both source and target cluster properties. 
   - These tests need to be migrated to use the new target cluster info format, which explicitly specifies target cluster shape and Spark properties.
   -  Till then, all new Profiling AutoTuner tests should be added in `ProfilingAutoTunerSuiteV2`

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/tuning/AutoTuner.scala:
```diff
@@ -316,13 +316,27 @@ class AutoTuner(
+  /**
+   * Kept for backwards compatibility.
+   */
+  private lazy val getAllSourceProperties: Map[String, String] = {
```
Could you explain in detail why the need to mention backward compatibility here ?

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/tuning/AutoTuner.scala:
```diff
@@ -470,6 +494,54 @@ class AutoTuner(
+   *
+   * TODO: To be removed in the future https://github.com/NVIDIA/spark-rapids-tools/issues/1710
+   */
+  private lazy val recommendedMaxBytesInFlight: Long = {
```
Checked the issue, this mentioned that we can get rid of this property. Then why are we adding this in the PR >

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/tuning/AutoTuner.scala:
```diff
@@ -531,69 +603,85 @@ class AutoTuner(
-      Right((pinnedMem, executorMemOverhead, executorHeap, setMaxBytesInFlight))
+      // Handle the case when the final executor memory overhead is larger than the
+      // available memory left for the executor.
+      if (execMemLeft < finalExecutorMemOverhead) {
```
Could you explain this a bit in detail. As to why is this added as part of this PR ? Was there a specific example that brought this out ?

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/tuning/AutoTuner.scala:
```diff
@@ -1246,11 +1337,16 @@ class AutoTuner(
       }
+
+      // Add user enforced recommendations
+      platform.userEnforcedRecommendations.foreach {
```
Have a general doubt, please correct me if I missed something.
So overall we have created a setting wherein first we check if the user has enforced some sparkProperty, if yes we use that. If not, we use the properties that we think make sense (across any property). Since that logic is enforced overall, why do we go over the userEnforcedRecommendation again and then append them again ( is this to cover what we might have missed ?). I observed this is rendered from the same set of properties from target cluster.
If that is the case, why not change the logic only at the end where we see if the user has enforced a property, we just use that to override what we might have appended. What detail am I missing here that needs a change this big and per conf validation ?

parthosa replied:
> why do we go over the userEnforcedRecommendation again and then append them again ( is this to cover what we might have missed ?) 

Simplified the logic to include the enforced recommendations during `initRecommendations()`

> What detail am I missing here that needs a change this big and per conf validation ?

The changes in this PR are mainly related to the interaction between the user enforced properties and how it would affect the recommendation logic for other properties or how it would affect the recommended cluster shape. 

For example: 
- User enforced value of memoryOverhead can be too large that might exceed the available memory limits. 
    - See UTs added in `ProfilingAutoTunerSuiteV2` for these cases.
- User enforced value of `spark.executor.instances` might affect the cluster shape. 
    - See UTs added in `ClusterRecommendationSuite` for these cases.

parthosa replied:
> Just for my understanding, would affect the cluster shape as in - it would affect the cluster shape we recommend to the user

Yes

> I am assuming a large part of users enforcing their properties would know what they are doing and since these properties are bound to the --target_worker_cluster , how does the enforcement logic tie to the ClusterRecommendation

The enforcement logic would dictate the recommended cluster shape. For example, enforced value of `spark.executor.instances` dictates how many GPUs and thus how many worker nodes to be recommended.

sayedbilalbari reviewed (commented):
Thanks @parthosa . I have a few doubts. 
It took me a while to understand the flow of AutoTuner. Would be great if we can have a place that documents the flow ->
- first we get cluster specific
- then we get memory settings
Something like this. 
With all the latest changes, it has become somewhat complicated to track how a change would evolve in the future or if it is something particular to a use case.

amahussein reviewed (commented):
Thanks @parthosa !

QQ:

- It might be confusing to have `preserved` option. For instance, the customer does not which values are going toe be modified by the AutoTuner before starting. Some users might dump everything in the preserved option.
- If the customer adds something that is not listed in the `tuningTable.yaml`. What happens in that case? is it going to be correctly processed and added to the bootstrao.conf.
- I see that the comments show some kline with "it was enforced by user". Later we should consider changingthe comments format to be more like a structured format. If it is not json file, then it should be some sort of a table "`name|old|new|policy(enforces/calculated/)|comment1"

parthosa commented:
> It might be confusing to have preserved option. For instance, the customer does not which values are going toe be modified by the AutoTuner before starting. Some users might dump everything in the preserved option.

We intend these options to be used only for advanced users who know their requirements.

> If the customer adds something that is not listed in the tuningTable.yaml. What happens in that case? is it going to be correctly processed and added to the bootstrao.conf.

Yes, they will be correctly processed added to the recommendation. I have added UT in `ClusterRecommendationSuite` where the user has provided `spark.memory.offHeap.size` (which is not listed in tuningTable) and it gets added in the recommendation.

Also note - This feature is only for Profiling Tool. Qual Tool does not support the target properties yet (#1647)

> I see that the comments show some kline with "it was enforced by user". Later we should consider changingthe comments format to be more like a structured format. If it is not json file, then it should be some sort of a table "name|old|new|policy(enforced/calculated/)|comment"

That is a good idea. Instead of dumping the comments as string, a structured format could be more useful. Filed #1717 for this.

sayedbilalbari commented on core/src/test/scala/com/nvidia/spark/rapids/tool/tuning/ProfilingAutoTunerSuiteV2.scala:
```diff
@@ -0,0 +1,317 @@
+          |- ${ProfilingAutoTunerConfigsProvider.notEnoughMemCommentForKey("spark.executor.memory")}
+          |- ${ProfilingAutoTunerConfigsProvider.notEnoughMemCommentForKey("spark.executor.memoryOverhead")}
+          |- ${ProfilingAutoTunerConfigsProvider.notEnoughMemCommentForKey("spark.rapids.memory.pinnedPool.size")}
+          |- ${ProfilingAutoTunerConfigsProvider.classPathComments("rapids.jars.missing")}
```
Same as above

parthosa replied:
For memory-related configurations, since we can verify whether the total allocated memory exceeds the limits, we will generate a warning message as:

```
- This node/worker configuration is not ideal for using the RAPIDS Accelerator
  for Apache Spark because it doesn't have enough memory for the executors.
  We recommend either using nodes with more memory or reducing 'spark.memory.offHeap.size',
  as off-heap memory is unused by the RAPIDS Accelerator, unless explicitly required by
  the application. Need at least 79360 MB memory per executor.
```

sayedbilalbari reviewed (changes requested):
Thanks @parthosa . Had a few comments

sayedbilalbari commented on core/src/test/scala/com/nvidia/spark/rapids/tool/tuning/ProfilingAutoTunerSuiteV2.scala:
```diff
@@ -0,0 +1,317 @@
+          |- 'spark.rapids.shuffle.multiThreaded.reader.threads' was not set.
+          |- 'spark.rapids.shuffle.multiThreaded.writer.threads' was not set.
+          |- 'spark.rapids.sql.batchSizeBytes' was not set.
+          |- ${ProfilingAutoTunerConfigsProvider.getEnforcedPropertyComment("spark.rapids.sql.concurrentGpuTasks")}
```
Mainly what happens in the case of a property with a user overriding them ,the overridden property being not optimal. Do we still use the same messaging as 'user enforced' ?

parthosa replied:
> Mainly what happens in the case of a property with a user overriding them ,the overridden property being not optimal. Do we still use the same messaging as 'user enforced' ?

Yes. User-enforced properties are for advanced users who have specific requirements and they will be applied even if the value is non-optimal.  However, AutoTuner would add a comment indicating that this was not AutoTuner's recommendation but user-enforced:
```
- 'spark.rapids.sql.concurrentGpuTasks' was user-enforced in the target cluster properties.
```

sayedbilalbari commented:
Thanks @parthosa . Let's wait on Ahmed to get back on this PR.

sayedbilalbari reviewed (approved):
LGTM !
